src/ra_log_segment_writer.erl

```diff
@@ -529,3 +532,26 @@ maybe_wait_for_segment_writer(SegWriter, TimeRemaining)
 maybe_wait_for_segment_writer(_SegWriter, _TimeRemaining) ->
     ok.
 
+maybe_upgrade_segment_file_names(System, DataDir) ->
+    Marker = filename:join(DataDir, ?UPGRADE_MARKER),
+    case ra_lib:is_file(Marker) of
+        false ->
+            ?INFO("segment_writer: upgrading segment file names to "
+                  "new format in dirctory ~ts",
+                  [DataDir]),
+            [begin
+                 Dir = filename:join(DataDir, UId),
+                 case file:list_dir(Dir) of
+                     {ok, Files} ->
+                         [ra_lib:zpad_upgrade(Dir, F, ".segment")
+                          || F <- Files, filename:extension(F) =:= ".segment"];
+                     {error, enoent} ->
+                         ok
+                 end
+             end || {_, UId} <- ra_directory:list_registered(System)],
```

Comment:
Question to check my understanding: one `ra_log_segment_writer` is spawned per system, right? So it should be safe to upgrade everything in the segment writer's system?

Reply:
correct, also it is started before any servers or even the wal